Fix bug with grow/shrink and wrong layout in the ZeroAwareAllocator

fitzgen opened the pull request:
This fixes a bug where we could accidentally pass the wrong `Layout` to the
underlying allocator when growing or shrinking an allocation. When we have an
already-zeroed block of size N, we can satisfy allocations of size `M` where `M
< N` if we are willing to accept some fragmentation slop. We currently do this
but only when the slop is relatively small (1/8 the size of the allocation or
less). If we handed out such an allocation, and then the user wants to grow or
shrink it, then we would pass the user's layout to the underlying allocator,
rather than the original layout. Passing the wrong layout to an allocator is a
violation of the `Allocator` interface's safety contract, and could therefore
lead to UB. In fact, this was leading to assertion failures when paired with
`dlmalloc`.

This commit also sets up and enables fuzzing via libfuzzer with our existing
mutator and oracle; previously we were only using the mutator and oracle for
property-testing style tests. I forgot to do this originally and just stopped
after I had the property-tests working with the mutator and oracle.

fitzgen committed:
Fix bug with grow/shrink and wrong layout in the `ZeroAwareAllocator`

This fixes a bug where we could accidentally pass the wrong `Layout` to the
underlying allocator when growing or shrinking an allocation. When we have an
already-zeroed block of size N, we can satisfy allocations of size `M` where `M
< N` if we are willing to accept some fragmentation slop. We currently do this
but only when the slop is relatively small (1/8 the size of the allocation or
less). If we handed out such an allocation, and then the user wants to grow or
shrink it, then we would pass the user's layout to the underlying allocator,
rather than the original layout. Passing the wrong layout to an allocator is a
violation of the `Allocator` interface's safety contract, and could therefore
lead to UB. In fact, this was leading to assertion failures when paired with
`dlmalloc`.

This commit also sets up and enables fuzzing via libfuzzer with our existing
mutator and oracle; previously we were only using the mutator and oracle for
property-testing style tests. I forgot to do this originally and just stopped
after I had the property-tests working with the mutator and oracle.

fitzgen committed:
Build and run the fuzzer in CI

fitzgen commented:
@cfallin mind giving this a once over?

cfallin commented on .github/workflows/rust.yml:
```diff
@@ -38,3 +39,14 @@ jobs:
+    - run: rustup toolchain install nightly
+    - run: rustup default nightly
+    - run: cargo install cargo-fuzz
+    - run: cargo fuzz build ops --sanitizer none
```
Would there be any value to enabling ASan here (i.e., omitting `--sanitizer none`) since we're testing an allocator with unsafe layout bits?

fitzgen replied:
I can investigate doing this in a follow up. My experience has been that things can be funky when ASan doesn't know about your custom allocator methods, but I'll try to get it up and running.

cfallin commented on crates/fuzzing/src/lib.rs:
```diff
@@ -46,9 +48,32 @@ impl Layout {
+        {
+            self.size
+        } else {
+            align
```
Could we have a comment here describing why we return just `align`? I'm not sure I fully grok it otherwise: if size overflows when trying to round up to alignment, we return just one alignment worth of size? Why wouldn't we panic (due to overflow) in that case?

cfallin replied:
(Is this to address a case found by fuzzing? I notice `max_size_for_align` bounds the size at `isize::MAX + 1 - align` and I suspect we want instead `isize::MAX - align` so there is always an alignment to round up to?)

fitzgen replied:
It's because, while we will only generate valid layouts ourselves, libfuzzer will mutate the encoded bytes arbitrarily and might create something that will decode into an invalid layout, so we have to have some kind of default for when it is invalid. I just just `align` arbitrarily.

Will add a comment.

cfallin replied:
Would it be possible to have a sort of validation pass after decoding and before propagating layouts into the API calls during fuzzing?

The concern I have is that this behavior is also user-facing, i.e. we might be silently masking overflow issues coming from user allocations (with enormously large allocation requests, but still).

fitzgen-f5 replied:
It is actually part of the `Allocator` interface's safety contract that `round_up(size, align) < isize::MAX` so that would be a bug on the caller side anyways. Given that, I'm not sure that any further changes here would actually improve the situation.

cfallin replied:
Ah! OK, that makes sense; can we add that to a comment too? I hadn't realized it was forbidden by contract.

cfallin commented on src/zero_aware_allocator.rs:
```diff
@@ -237,34 +275,126 @@ where
+        let zeroed = &mut *zeroed;
+        let node = match zeroed.live_set.remove(&ptr) {
+            Some(node) => {
+                // The actual layout should imply the user layout.
```
s/imply/contain/ maybe?

cfallin commented on src/zero_aware_allocator.rs:
```diff
@@ -282,18 +412,49 @@ where
     #[inline]
     fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        if layout.size() == 0 {
+            let dangling = core::ptr::without_provenance::<u8>(layout.align());
```
Could we add a comment here describing what we're returning? It certainly looks a little scary otherwise :-)

Something like: "the return value for a zero-sized allocation can be an arbitrary pointer, but we must ensure it is non-null. Thus, we invent a pointer whose numeric value is equal to the alignment (i.e., the first possible non-null pointer of that alignment)."

And then note that the deallocation receives the layout as well and has a corresponding case for zero size.

cfallin replied:
Finally: could we note why we're handling this explicitly here (vs. delegating to the inner allocator)? Is it because we want to ensure our liveset tracking isn't exposed to this?

fitzgen replied:
> Is it because we want to ensure our liveset tracking isn't exposed to this?

Yeah, we just want to make sure we don't ever allocate/queary bookkeeping metadata for zero-sized allocations that won't ever need it. In practice, LLVM should inline call sites, where we will almost always have a constant `Layout`, and these checks should all get cleaned up and removed.

cfallin commented on src/zero_aware_allocator.rs:
```diff
@@ -302,28 +463,87 @@ where
+            PreGrow::Reuse {
+                ptr,
+                actual_layout: _,
+            } => return Ok(ptr),
```
Can we add a note justifying why we don't need to update any metadata here -- basically, the caller is responsible for always knowing the current layout, and the size of the actual block didn't change, and we only track the actual block size, so we don't need to update any bookkeeping?

cfallin commented on src/zero_aware_allocator.rs:
```diff
@@ -302,28 +463,87 @@ where
+                // NB: we cannot assume that the fragmentation slop (if any) is
+                // still zeroed, because we give out the whole block as a
+                // `NonNull<[u8]>` slice to the user. Therefore, we have to
+                // manually zero here.
```
Interesting -- is there a reason we have to return the full block as a slice rather than only up to the user layout's size?

fitzgen replied:
It allows callers to make use of the full buffer, e.g. a `Vec` can use that extra capacity, delaying how long until it needs to realloc and double its capacity. Our embedding of this crate will also use this feature shortly.

cfallin replied:
If the growth within this gap (between the requested and the actual available) is relatively cheap anyway, per this crate's implementation, and if handing out in a more restricted way lets us avoid some zeroing, I wonder if the tradeoff might be better to hold back that extra part?

(In other words: we're handing out extra for efficiency's sake, to defer a need for growth; but in doing so we're creating extra work because we lose track of what is zeroed, so maybe that efficiency gain is lost?)

fitzgen replied:
If the caller makes use of the extra capacity, they won't ever call `grow` within that extra capacity and they won't ever actually hit this code path. This is the case for our embedding, fwiw.

If we don't expose the extra capacity to the caller, then they are incentivized to do doubling-growth to amortize cost, because they cannot know that a small grow *might* be free in a particular case, which means that the new size of subsequent `grow[_zeroed]` calls definitely won't fit within this extra capacity, and so the extra capacity never gets used in practice.

So the trade off is a potential slow(er) path that does up to 1/8 * size zeroing *iff* you don't use the extra capacity and do a very small `grow` versus more fragmentation combined with larger `{allocate,grow}_zeroed` calls happening sooner / more frequently. It seems to me like the rare zeroing is the better option.

cfallin replied:
Hmm, yeah, I suppose with no way to communicate the "there's a little extra space cheaply if you ask for it" state, it does seem better to grant the full size right away. Fair enough!

cfallin commented on src/zero_aware_allocator.rs:
```diff
@@ -354,8 +574,35 @@ where
+
+        // Make sure that we are passing the original, underlying `Layout` to
+        // the inner allocator. After shrinking, this pointer will no longer
+        // managed by us, so remove and deallocate its block-info node from the
```
s/will no longer managed/will no longer be managed/

cfallin reviewed (approved):
This generally looks good to me -- thanks for working this out so carefully/thoroughly! It's too bad we need additional bookkeeping for the partially-used blocks but I don't see a better way either.

Comments below are *mostly* asking for clarification/justification comments...

fitzgen committed:
Clean up zero-sized allocation handling in `ZeroAwareAllocator`

Always defer to the inner allocator for these; assert that we are never dealing
with zero-sized allocations in `[de]allocate_already_zeroed`.

fitzgen committed:
Document size/align getters for fuzzing infra; saturate rather than choose `align` as size on raw size being too big

fitzgen committed:
Add some more comments from review

fitzgen committed:
Add some more comments

fitzgen committed:
Move fuzz CI job to test theory about CI
